Automattic/jetpack
# Connection Error Handling

This PR introduces an Error Handling class for the connection package. This class will catch errors when wpcom makes an xml-rpc request to the site and the request signature is invalid.

The goal is to give better feedback to the user on how to solve the problem and, when possible, self-heal.

## How it works

When an XML-RPC request with a broken signature is received, Jetpack does not respond with any useful information (for security reasons). This leaves the user and WP.COM with no information about what is broken in the connection.

This PR adds a workflow where the site with a broken connection can securely confirm the error and inform it both to WP.COM and to the user.

1. It all starts with a request to the XML-RPC endpoint. When the signature is invalid:

2. checks if the error_code triggered was triggered in the last hour. It will only process it once an hour to avoid overloads

3. Stores the error in the database as "unverified", because we don't know if this broken request actually came from WP.COM

4. Creates a nonce for this new error.

5. Encrypts the error and send it over to WP.COM

6. WP.COM receives it and will also only process the request every once in a while. It then checks it against its database and, if the error looks legit (it has the same tokens present in WP.COM database) it will send a new request to the site, informing the token generated in step 4

7. The site receives the request from WP.COM in this new REST endpoint introduced by this PR. It checks the token and "verify" the error, meaning it will now be in the list of verified errors.

8. If the error does not get verified after 3 days, it gets deleted

9. Now that the error is verified, each error code will trigger a handle in initialization that can do different things. This PR only treats a certain number of error codes that indicates the `blog_token` is broken. It will display a simple `admin_notice` with an error message.

![captura-de-tela-2020-06-19-acc80s-17 29 53](https://user-images.githubusercontent.com/971483/85784397-9f5f6900-b6fe-11ea-9747-910032368a3b.png)

#### Jetpack product discussion
p1HpG7-9q9-p2

#### Does this pull request change what data or activity we track or use?


Yes, it collects details about connection errors (broken tokens, etc)

## Testing instructions:






To test this PR, we're going to use the "broken tokens" plugin available in the `mu-plugins` folder of this repository (which works out of the box in our docker environment)

* Start with a site successfully connected to wp.com
* Go to Jetpack > Broken tokens
* Click on "Store these options" to save your valid tokens (you will need this to fix your connection later)
* Now choose an option and break the connection. For a start, choose "Set invalid blog token"
* Go to Jetpack > XML-RPC Errors
* There should not be any errors there yet. If you see any errors, delete them before you start.
* Click on the "Jetpack Debugger" link and go to the jetpack debugger. It should show an error for your site
* Go back to Jetpack > XML-RPC Errors
* Now you should see an error in the Unverified Errors list
* Check the error nonce and timestamp. copy and save it somewhere
* Reload Jetpack Debugger. It will display the same error again
* Reload XML-RPC errors page and verify that the timestamp and nonce did not change. This verifies that the gate worked. It did not process the same error twice within an hour
* (note, `JETPACK_DEV_DEBUG` must be false for this to work)

At this point, the error was sent to WP.COM and will be processed. If confirmed WP.COM will trigger a request to our new API endpoint.

For our tests, we don't need to wait. 

* Click the "Verify Error (via API)" button
* Check that the error is copied over to the Verified errors list
* Now navigate the WP admin dashboard and confirm there is an admin notice at the top of the screen informing about the connection error.
* Disconnect your site
* Confirm that the error message is gone

## Complementary tests

* Install the [Client Example](https://github.com/Automattic/client-example) plugin and test it with Jetpack deactivated (you might need to copy the connection package over to the client example `vendor/automattic` folder if symlink is not in place)
* test with different errors (breaking the connection in different ways)
** notice the admin notice will only appear for blog token errors, which means when the requested user id is 0 or malformed and the error code is one of the following:

```
case 'malformed_token':
'token_malformed':
'no_possible_tokens':
'no_valid_token':
'unknown_token':
'could_not_sign':
'invalid_token':
'could_not_sign':
'token_mismatch':
'invalid_signature':
'signature_mismatch':
```

#### Proposed changelog entry for your changes:

* adds xml-rpc error handling to the connection package
